### PR TITLE
Add functionality for editing Cody messages

### LINKED ISSUE #50974: Add ability to edit and resubmit chat message
I want the same "edit and resubmit" functionality that ChatGPT has. I've been missing this a lot.

Sometimes I come up with an elaborate (and super smart, I swear) prompt and then notice I forgot something. But I can't get to the Markdown of the prompt so I essentially need to retype it somewhere else and resubmit.

### PR DESCRIPTION
**Pull request by SuperAuguste**
Closes #50974
## Test plan

Tested locally in VSCode and it *seems* to work - feel free to suggest additional testing methodologies :)

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.01% (+1.27 kb) | 0.01% (+1.27 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/auguste/cody-edit-message/compare-report.html;tab=live_object) for a full comparison between the commits 7e3eaae and 629739e or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/cody-ui/src/chat/Transcript.tsx [hunk 1]
```diff
@@ -4,7 +4,7 @@ import classNames from 'classnames'
 
 import { ChatMessage } from '@sourcegraph/cody-shared/src/chat/transcript/messages'
 
-import { FeedbackButtonsProps } from '../Chat'
+import { ChatUITextAreaProps, EditButtonAction, EditButtonProps, FeedbackButtonsProps } from '../Chat'
```
Comment (abeatrix) [inline client/cody-ui/src/chat/Transcript.tsx:7]:
```suggestion
import { ChatUITextAreaProps, EditButtonProps, FeedbackButtonsProps } from '../Chat'
```

**Review by abeatrix — CHANGES_REQUESTED** [review]
This is really really cool! So cool I went over a bit on the review so please let me know if my suggestions don't make sense 😆 
Most of them are just nits, except the new changes allow users to replace with an empty input so requesting a change on that 🙇‍♀️

**Comment by mrnugget** [comment]
I pushed two commits that should help with the ESLint errors

File: client/cody-ui/src/chat/TranscriptItem.tsx [hunk 2]
```diff
@@ -31,71 +31,141 @@ export const TranscriptItem: React.FunctionComponent<
...
+        TextArea && beingEdited && editButtonOnSubmit ? (
+            <TextArea
+                className={classNames(styles.chatInput)}
+                rows={5}
+                value={formInput}
+                autoFocus={true}
+                required={true}
+                onInput={({ target }) => {
+                    const { value } = target as HTMLInputElement
+                    inputHandler(value)
```
Comment (abeatrix) [inline client/cody-ui/src/chat/TranscriptItem.tsx:81]:
```suggestion
                    setFormInput(value)
```
nit: this way we can remove the `inputHandler` above

**Review by abeatrix — APPROVED** [review]
Nice! The update works great on my local vscode instance 👍

File: client/cody-ui/src/chat/TranscriptItem.tsx (hunk 2 above)
Comment (SuperAuguste) [reply client/cody-ui/src/chat/TranscriptItem.tsx:81]: Oops nice catch, copied this line-for-line from somewhere else in the codebase when it isn't necessary here 😅

**Comment by mrnugget** [comment]
@SuperAuguste I'm gonna go ahead and merge this. Thanks!

**Pull request merged**